src/controller/scan/base_controller.go

```diff
@@ -687,23 +713,111 @@ func (bc *basicController) launchScanJob(ctx context.Context, trackID string, ar
 	params[sca.JobParameterMimes] = mimes
 	params[sca.JobParameterRobot] = robotJSON
 
-	// Launch job
-	callbackURL, err := bc.config(configCoreInternalAddr)
-	if err != nil {
-		return "", errors.Wrap(err, "launch scan job")
-	}
-	hookURL := fmt.Sprintf("%s/service/notifications/jobs/scan/%s", callbackURL, trackID)
-
-	j := &jm.JobData{
+	j := &task.Job{
 		Name: job.ImageScanJob,
-		Metadata: &jm.JobMetadata{
+		Metadata: &job.Metadata{
 			JobKind: job.KindGeneric,
 		},
 		Parameters: params,
-		StatusHook: hookURL,
 	}
 
-	return bc.jc().SubmitJob(j)
+	extraAttrs := map[string]interface{}{
+		artifactIDKey:  artifact.ID,
+		robotIDKey:     robot.ID,
+		reportUUIDsKey: reportIDs,
+	}
+
+	_, err = bc.taskMgr.Create(ctx, executionID, j, extraAttrs)
+	return err
+}
+
+// listScanTasks returns the tasks of the reports
+func (bc *basicController) listScanTasks(ctx context.Context, reportUUIDs []string) ([]*task.Task, error) {
+	if len(reportUUIDs) == 0 {
+		return nil, nil
+	}
+
+	arr := make([]string, len(reportUUIDs))
+	for i, reportUUID := range reportUUIDs {
+		arr[i] = "'" + reportUUID + "'"
+	}
+
+	// condition looks like (extra_attrs->'report_uuids')::jsonb ?| array['uuid1', 'uuid2', 'uuid3']
+	// that condition means to select the task record when any element in the report_uuids of the extra_attr is in the set ('uuid1', 'uuid2', 'uuid3')
+	condition := fmt.Sprintf("(extra_attrs->'%s')::jsonb ?| array[%s]", reportUUIDsKey, strings.Join(arr, ","))
+
+	return bc.taskMgr.List(ctx, q.New(q.KeyWords{"extra_attrs_element": condition}))
+}
+
+func (bc *basicController) assembleReports(ctx context.Context, reports ...*scan.Report) error {
+	reportUUIDs := make([]string, len(reports))
+	for i, report := range reports {
+		reportUUIDs[i] = report.UUID
+	}
+
+	tasks, err := bc.listScanTasks(ctx, reportUUIDs)
+	if err != nil {
+		return err
+	}
+
+	reportUUIDToTasks := map[string]*task.Task{}
+	for _, task := range tasks {
+		for _, reportUUID := range getReportUUIDs(task.ExtraAttrs) {
+			reportUUIDToTasks[reportUUID] = task
+		}
+	}
+
+	for _, report := range reports {
+		if task, ok := reportUUIDToTasks[report.UUID]; ok {
+			report.Status = task.Status
+			report.StartTime = task.StartTime
+			report.EndTime = task.EndTime
+		} else {
+			report.Status = job.ErrorStatus.String()
+		}
+	}
+
+	return nil
+}
+
+func getArtifactID(extraAttrs map[string]interface{}) int64 {
+	var artifactID float64
+	if extraAttrs != nil {
+		if v, ok := extraAttrs[artifactIDKey]; ok {
+			artifactID, _ = v.(float64) // int64 Unmarshal to float64
```

Comment:
err?

Reply:
This value must be the float64, so err does not need here.